topdown/crypto.go

```diff
@@ -378,42 +417,69 @@ func getX509CertsFromPem(pemBlocks []byte) ([]*x509.Certificate, error) {
 	return x509.ParseCertificates(decodedCerts)
 }
 
-func getRSAPrivateKeyFromString(key string) (interface{}, error) {
-	// if the input is PEM handle that
+func getPrivateKeyFromString(key string) (interface{}, error) {
 	if strings.HasPrefix(key, "-----BEGIN") {
-		return getRSAPrivateKeyFromPEM([]byte(key))
+		return getPrivateKeyFromPem(key)
 	}
 
-	// assume input is base64 if not PEM
 	b64, err := base64.StdEncoding.DecodeString(key)
 	if err != nil {
 		return nil, err
 	}
-
-	return getRSAPrivateKeyFromPEM(b64)
+	return getPrivateKeyFromPem(string(b64))
 }
 
-func getRSAPrivateKeyFromPEM(pemBlocks []byte) (interface{}, error) {
-
-	// decode the pem into the Block struct
-	p, _ := pem.Decode(pemBlocks)
-	if p == nil {
-		return nil, fmt.Errorf("failed to parse PEM block containing the key")
+func getPrivateKeyFromPem(pemBlocks string) (interface{}, error) {
+
+	var validPrivateKeys []crypto.PrivateKey
+
+	if len(pemBlocks) != 0 {
+
+		bs := []byte(pemBlocks)
+		for len(bs) > 0 {
```

Comment:
This is great, it'd be nice to exercise this with a test case of varied PEM input like this: https://go.dev/play/p/2XqZn1wbgC2 so we can see that it extracts the expected Private Key blocks and ignores the others as we need.